Polygon Mesh Processing:  Add envelope test (#5090)
CGAL/cgal: Polygon_mesh_processing/doc/Polygon_mesh_processing/Polygon_mesh_processing.txt

\subsubsection InsideExample Inside Test Example
\cgalExample{Polygon_mesh_processing/point_inside_example.cpp}

\subsection PMPEnvelope Polyhedral Envelope Containment Check

> The class `CGAL::Envelope` provides functors that test whether a query point, segment or triangle

Reviewer: ```suggestion
The class `CGAL::Envelope` provides functors that test whether a query point, segment, or triangle
```

> is fully contained in a <em>polyhedral envelope</em> of a triangle mesh.

Reviewer: works with triangle soups too, right?

Author: yes, but that code is not yet written.

Reviewer: Will you allow them in this submission?

Author: no.

Reviewer: I added soup, please update.

\cgalFigureBegin{envelopeFig, envelope.png}
Two polyhedral envelopes.
\cgalFigureEnd

While the user provides a distance

> the envelope is not the Minkowski sum with a sphere, because this would have cylindrical and spherical

Reviewer: radius of the sphere = distance

Reviewer: I would not say anything about convex edges but that the MS is the union of cylinders, sphere and prisms.

Author: I find it  instructive to see where the cylinders and spheres would come from.

patches at convex edges and vertices of the input triangle mesh. Instead, the envelope consists
of a collection of prisms that are guaranteed to be inside the Minkowski sum envelope, hence the
term polyhedral envelope.  This containment test is exact for the polyhedral envelope, and
conservative in the sense that, if a query is inside the polyhedral envelope, we can be sure

> that it is also in the Minkowski sum envelope, but if it is outside the polyhedral envelop

Reviewer: ```suggestion
that it is also in the Minkowski sum envelope, but if it is outside the polyhedral envelope
```

we do not know where it is with respect to the Minkowski sum envelope.

Such a test is used by the class `Surface_mesh_simplification::Envelope_filter` of the

> package \refPkgSurfaceMeshSimplification, in order to simplify respecting a tolerance.

Reviewer: ```suggestion
package \ref PkgSurfaceMeshSimplification, in order to simplify respecting a tolerance.
```

> package \refPkgSurfaceMeshSimplification, in order to simplify respecting a tolerance.

Reviewer: ```suggestion
package \refPkgSurfaceMeshSimplification, in order to simplify a triangle mesh while respecting an approximation tolerance.
```

\subsubsection InsideExample Polyhedral Envelope Example
\cgalExample{Polygon_mesh_processing/polyhedral_envelope.cpp}

> The algorithm is described in \cgalCite{cgal:Wwshap-eepec-20} and can be summarized as follows.

Reviewer: Why is this section in example?

Author: Because other features such as the visitor or the bounded volume are also only in examples.

> First for each triangle a <em>prism</em> is constructed with an offset of \f$ \epsilon/\sqrt(3) \f$ where \f$ \epsilon\f$ is the tolerance,

Reviewer: tolerance vs. distance

Reviewer: ping

> and all prisms are stored in an AABB tree.  For a query point the algorithm checks if is inside a prism.

Reviewer: ```suggestion
and all prisms are stored in an AABB tree.  For a query point the algorithm checks if it is inside a prism.
```

For a query segment or triangle the algorithms checks if the query is completely covered.

> The details of how to check this covering can be found in the paper of Wang et al.

Reviewer: ```suggestion
The details of how to check this covering can be found in \cgalCite{cgal:Wwshap-eepec-20}.
```

Author: We have that \cite already in the first sentence of the paragraph.

Author: I moved the \cite to the first sentence.

\subsubsection InsideExample Inside Test Example
\cgalExample{Polygon_mesh_processing/point_inside_example.cpp}

\subsection PMPEnvelope Polyhedral Envelope Containment Check

The class `CGAL::Envelope` provides functors that enable you to check if a query point, segment, or triangle
is fully contained in a <em>polyhedral envelope</em> of a triangle mesh.

> Given a distance \f$ \delta \f$ we compute for each triangle  a <em>prism</em> by intersecting two halfspaces

Reviewer: delta vs epsilon (refman)

Reviewer: You are not given delta but epsilon, right?

such that they contain the triangle.

The <em>polyhedral envelope</em> with a tolerance \f$ \epsilon \f$ then is the union of the prisms of all triangular faces
with  \f$ \delta = \epsilon / \sqrt(3) \f$

\cgalFigureBegin{envelopeFig, envelope.png}
The prism for a single triangle, the polyhedral envelope as well as the Minkowski sum envelope for a triangle mesh.
\cgalFigureEnd

For containent checks the polyhedral envelope is a conservative approximation of the Minkowski sum with a sphere of radius \f$ \epsilon \f$.
The latter has cylindrical and spherical patches at convex edges and vertices of the input triangle mesh.

The polyhedral envelope is guaranteed to be inside the Minkowski sum envelope.
This containment test is exact for the polyhedral envelope, and conservative for the Minkowswi sum envelope:
If a query is inside the polyhedral envelope, we can be sure
that it is also in the Minkowski sum envelope, but if it is outside the polyhedral envelope
we do not know where it is with respect to the Minkowski sum envelope.

Such a test is used by the class `Surface_mesh_simplification::Envelope_filter` of the

> package \refPkgSurfaceMeshSimplification, in order to simplify within  a given tolerance.

Reviewer: ```suggestion
package \refPkgSurfaceMeshSimplification, in order to simplify within a given tolerance.
```

For containent checks the polyhedral envelope is a conservative approximation of the Minkowski sum with a sphere of radius \f$ \epsilon \f$.
The latter has cylindrical and spherical patches at convex edges and vertices of the input triangle mesh.

The polyhedral envelope is guaranteed to be inside the Minkowski sum envelope.
This containment test is exact for the polyhedral envelope, and conservative for the Minkowswi sum envelope:
If a query is inside the polyhedral envelope, we can be sure
that it is also in the Minkowski sum envelope, but if it is outside the polyhedral envelope
we do not know where it is with respect to the Minkowski sum envelope.

The algorithm of  Wang et al \cgalCite{cgal:Wwshap-eepec-20} for the polyhedral envelope containment check
can be summarized as follows.
The prisms of the faces of the triangle mesh are stored in an AABB tree, which is used to quickly
identify the prisms whose bounding box overlaps with the query.

For a query point the algorithm checks if it is inside one of these prisms.
For a query segment or triangle the algorithms checks if the query is completely covered.
The details of how to check this covering can be found in the paper.

> The polyedral envelope containment check is used by the class `Surface_mesh_simplification::Envelope_filter`

Reviewer: ```suggestion
The polyhedral envelope containment check is used by the class `Surface_mesh_simplification::Envelope_filter`
```

by Konstantinos Katrioplas, under supervision of Jane Tournois. It was finalized by Mael Rouxel-Labbé and
integrated in \cgal 5.0.

Functionalities related to mesh and polygon soup repair have been introduced steadily over multiple versions
since \cgal 4.10, in joint work between Sébastien Loriot and Mael Rouxel-Labbé.

> The polyhedral envelope containment check was integrated in \cgal 5.2. The implementation makes use of the

Reviewer: ```suggestion
The polyhedral envelope containment check was integrated in \cgal 5.3. The implementation makes use of the
```
